packages/eslint-plugin/docs/rules/no-empty-object-type.mdx

```diff
@@ -0,0 +1,74 @@
+---
+description: 'Disallow accidentally using the "empty object" type.'
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-empty-object-type** for documentation.
+
+The `{}`, or "empty object" type in TypeScript is a common source of confusion for developers unfamiliar with TypeScript's structural typing.
+`{}` represents any _non-nullish value_, including literals like `0` and `""`:
+
+```ts
+let anyNonNullishValue: {} = 'Intentionally allowed by TypeScript.';
+```
+
+Often, developers writing `{}` actually mean either:
+
+- `object`: representing any _object_ value
+- `unknown`: representing any value at all, including `null` and `undefined`
+- An object with no properties: which can't easily be represented in TypeScript's structural type system
+
+In other words, the "empty object" type `{}` really means _"any value that is defined"_.
+That includes arrays, class instances, functions, and primitives such as `string` and `symbol`.
+
+To avoid confusion around the `{}` type allowing any _non-nullish value_, this rule bans usage of the `{}` type.
+
+:::tip
+If you do have a use case for an API allowing `{}`, you can always use an [ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) or [disable the rule in your ESLint config](https://eslint.org/docs/latest/use/configure/rules#using-configuration-files-1).
+:::
+
+Note that this rule does not report:
+
+- `{}` as a type constituent in an intersection type (e.g. `type NonNullable<T> = T & unknown`), as this is useful in type system operations.
+- `{}` as an empty function body with a `void` return type (e.g. `() => {}`), as this is a function body, not a type.
+
+## Examples
+
+<Tabs>
+<TabItem value="❌ Incorrect">
+
+```ts
+let anyObject: {};
+let anyValue: {};
+let emptyObject: {};
+```
+
+</TabItem>
+<TabItem value="✅ Correct">
+
+```ts
+let anyObject: object;
+let anyValue: unknown;
+let emptyObject: Record<string, never>;
+```
+
+</TabItem>
+</Tabs>
+
+## When Not To Use It
+
+If your code commonly needs to represent the _"any non-nullish value"_ type, this rule may not be for you.
+Projects that extensively use type operations such as conditional types and mapped types oftentimes benefit from disabling this rule.
+
+## Further Reading
+
+- [Enhancement: [ban-types] Split the {} ban into a separate, better-phrased rule](https://github.com/typescript-eslint/typescript-eslint/issues/8700)
+- [The Empty Object Type in TypeScript](https://www.totaltypescript.com/the-empty-object-type-in-typescript)
+
+## Related To
+
+- [`no-empty-interface`](./no-empty-interface.mdx)
```

Comment:
I feel like these two rules are really one, maybe as an option `allowInterface` for declaration merging

Reply:
Hmm, I see them as separate rules. They're definitely related but I think the justifications for disabling either are very different.

Reply:
Interesting. So just confirming, you're suggesting we should merge `no-empty-interface` into this rule for v8, too?

Reply:
Yes, exactly.

Reply:
Slept on it and ruminated - I'm ever so slightly leaning away from wanting to merge. But can't really make up my mind. 🫠

On one side, I do like that they're essentially targeting the same motivation, as you noted: that any non-nullable value is assignable to the `{}` shape. So it does make sense to me to put them together.

On the other hand, I think specifically the object type alias form is kind of controversial, and a lot of users will want to be able to disable `no-empty-object-type`. Maybe less so now that we're fixing the docs but I'm still guessing a non-trivial portion of users. I'm also thinking that those users likely won't want to disable `no-empty-interface`.

We could solve that by adding an `allowTypeAlias` option to the rule... but then we'd also have to mention that in the report message. Which IMO is getting to the point of a little more complication than is ideal. This is a big part of why I've started leaning towards smaller, less configurable lint rules. There might be more of them to lean but each gets much simpler to understand.

So my current proposal would be: leave separate in v8 for ideation, then maybe merge in v9 if we think it'd be valuable?

Reply:
the difference between the two is that `no-empty-interface` allows you to ban `interface Foo extends Bar {}` in favour of `type Foo = Bar`.
However a similar rule doesn't make sense for type aliases.

But there's no reason we couldn't merge them and keep the `allowSingleExtends` option just for interfaces.

but you're JoshG isn't wrong - the scope is different because the current form bans `{}` anywhere - not just in `type` aliases.

I personally can't think of a reason you'd ever want to have one banned, but not the other.

Reply:
no strong opinions on this!

Reply:
I will say, though, it does feel like yet another breaking change for users just to preserve the existing behavior. So, I guess "no strong opinion" defaults to not merging in this case 🤷

Reply:
> it does feel like yet another breaking change for users just to preserve the existing behavior

I would have said the opposite - it's a breaking change TO merge.
Cos if you don't merge - people are just switching from `ban-types` to this new rule.
But if you merge they ALSO need to switch from `no-empty-interface` to this new rule.